source/common/matcher/matcher.h

```diff
@@ -191,15 +198,22 @@ template <class DataType, class ActionFactoryContext> class MatchTreeFactory {
     }
     case MatcherType::MatcherTree::kPrefixMatchMap:
       NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
-    case MatcherType::MatcherTree::kCustomMatch:
-      NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+    case MatcherType::MatcherTree::kCustomMatch: {
+      auto& factory = Config::Utility::getAndCheckFactory<CustomMatcherFactory<DataType>>(
```

Comment:
It's the return type that requires handling any `DataType` in a function. It would be better done via overloads, not templates IMHO if we can define a type hierarchy for it. Basically, we have to provide some kind of polymorphism over the evaluation "context", which is DataType. Current design does it via factory registration. It's possible to do it via overloads within one factory but not sure what we gain from that since it is still static code registration for each matcher and datatype.